ext/fg/js/frame-offset-forwarder.js

```diff
@@ -91,6 +96,24 @@ class FrameOffsetForwarder {
         this._forwardFrameOffset(offset, uniqueId);
     }
 
+    _getOpenShadowRootSourceFrame(sourceWindow) {
```

Comment:
This method is likely to be very expensive, particularly on complex pages. Consider something like this which can skip many operations once the result is found:

```js
// use document.documentElement as startElement
function findFrameWithContentWindow(startElement, contentWindow) {
  const ELEMENT_NODE = Node.ELEMENT_NODE;
  const elements = [startElement];
  while (elements.length > 0) {
    const element = elements.shift();
    switch (element.nodeName.toUpperCase()) { // toUpperCase potentially costly
      case 'IFRAME':
      case 'FRAME':
        if (element.contentWindow === contentWindow) {
          return element;
        }
        break;
    }

    const shadowRoot = element.shadowRoot;
    if (shadowRoot) {
      for (const child of shadowRoot.children) {
        if (child.nodeType === ELEMENT_NODE) {
          elements.push(child);
        }
      }
    }
    
    for (const child of element.children) {
      if (child.nodeType === ELEMENT_NODE) {
        elements.push(child);
      }
    }
  }
}
```

Reply:
Yeah, it's a naive approach and basically what I ended up with after testing in the browser console. However, I think this is very rare in the wild, and you only end up there if a scan actually happens in an iframe that's in shadow dom (or potentially some other unknown case), plus it's throttled to max 1 per second. I can try to optimize it a bit to help with the worst case.

Reply:
https://github.com/FooSoft/yomichan/pull/460/commits/859c8480bd330502ed1ba91067c023c28139c5a1

I can also add some better data structure for the cache to avoid memory leaks if needed. I wonder if it would also need some error handling when the frames are removed from DOM.

Reply:
I doubt the cache will really be a significant cause for memory leaks unless some website is deliberately doing something stupid, but maybe a simple cache size limit could be used when adding a new item.

```js
function addSet(set, value, limit) {
  limit -= set.size;
  if (limit >= 0) {
    for (const value2 of set) {
      set.delete(value2); // This appears to be valid to do during iteration
      if (limit <= 0) { break; }
      --limit;
    }
  }
  set.add(value);
}
```

> I wonder if it would also need some error handling when the frames are removed from DOM.

Just check `document.documentElement.contains(element)` before returning, and probably remove from caches.

Reply:
https://github.com/FooSoft/yomichan/pull/460/commits/8b4d5473d33256f4e050c0f85af3fcd4479b8764

It's a nice property of `Set` that the elements preserve insertion order. I was thinking of creating a parallel `Array` to remove older items first.

I used `Node.isConnected` instead of `contains` because I believe that `contains` doesn't work for shadow DOM unless you know the root element. It required bumping the minimum Firefox version to 53, though.